tensorflow/c/kernels.cc

```diff
@@ -551,3 +554,257 @@ TF_Tensor* TF_AllocateTemp(TF_OpKernelContext* context, TF_DataType dtype,
+  return Status::OK();
+}
+
+tensorflow::Status PrepareToUpdateVariable(TF_OpKernelContext* ctx, tensorflow::Tensor* tensor,
+                               bool copy_on_read_mode,
+                               bool variantType,
+                               void (*copyFunc)(TF_OpKernelContext * ctx, TF_Tensor *source, TF_Tensor *dest)) {
+
+  using namespace tensorflow;
+  auto* context = reinterpret_cast<::tensorflow::OpKernelContext*>(ctx);
+  if (copy_on_read_mode || !tensor->RefCountIsOne()) {
+    // Tensor's buffer is in use by some read, so we need to copy before
+    // updating.
+    Tensor tmp;
+    if(variantType) {
+      AllocatorAttributes attr;
+      attr.set_on_host(true);
+      TF_RETURN_IF_ERROR(
+          context->allocate_temp(tensor->dtype(), tensor->shape(), &tmp, attr));
+
+      const auto elements_in = tensor->flat<Variant>();
+      auto elements_out = tmp.flat<Variant>();
+      for (int64 i = 0; i < elements_in.size(); ++i) {
+        elements_out(i) = elements_in(i);
+      }
+    } else {
+      AllocatorAttributes attr;
+      attr.set_gpu_compatible(true);
+      attr.set_nic_compatible(true);
+      TF_RETURN_IF_ERROR(
+          context->allocate_temp(tensor->dtype(), tensor->shape(), &tmp, attr));
+      tensorflow::Status s;
+      TF_Tensor *tf_tmp = TF_TensorFromTensor(tmp, &s);
+      TF_Tensor *tf_tensor = TF_TensorFromTensor(*tensor, &s);
+      copyFunc(ctx, tf_tensor, tf_tmp);
+    }
+    *tensor = tmp;
+  }
+  return Status::OK();
+}
+
+void TF_AssignVariable(TF_OpKernelContext* ctx,
+                       int input_index,
+                       int value_index,
+                       void (*copyFunc)(TF_OpKernelContext * ctx, TF_Tensor *source, TF_Tensor *dest),
+                       TF_Status* status) {
+  auto* cc_ctx = reinterpret_cast<::tensorflow::OpKernelContext*>(ctx);
+  tensorflow::AllocatorAttributes allocator_attr;
+  tensorflow::core::RefCountPtr<tensorflow::Var> variable;
+  const tensorflow::Tensor& value = cc_ctx->input(value_index);
+  auto c_stream = static_cast<stream_executor::CStream*>(
+      cc_ctx->op_device_context()->stream()->implementation());
+  OP_REQUIRES_OK(cc_ctx, LookupOrCreateResource<tensorflow::Var>(
+                                cc_ctx, HandleFromInput(cc_ctx, input_index), &variable,
+                                [&value](tensorflow::Var** ptr) {
+                                  *ptr = new tensorflow::Var(value.dtype());
+                                  *(*ptr)->tensor() = value;
+                                  (*ptr)->is_initialized = true;
+                                  return tensorflow::Status::OK();
+                                }));
+  tensorflow::mutex_lock ml(*variable->mu());
+
+  if (variable->copy_on_read_mode.load()) {
+    tensorflow::Tensor* tmp;
+    tensorflow::AllocatorAttributes attr;
+    attr.set_gpu_compatible(true);
+    attr.set_nic_compatible(true);
+    OP_REQUIRES_OK(cc_ctx,
```

Comment:
I  think the tensor in allocate_tmp should be tensor `tmp`, not `variable->tensor()`? `

Reply:
Good catch!,  updated.